test/src/691-metalink/main

```diff
@@ -0,0 +1,230 @@
+  #sudo systemctl daemon-reload
+  sudo cvmfs_server mount $CVMFS_TEST_REPO               || return 44
+
+  local internal2_log="${scratch_dir}/internal2.log"
+  echo "*** Start second HTTP server to serve internal files (logging to $internal2_log)"
+  CVMFS_TEST_691_INTERNAL2_PID="$(open_http_server $internal_storage2 $internal_port2 $internal2_log)"
+  [ ! -z $CVMFS_TEST_691_INTERNAL2_PID ] && kill -0 $CVMFS_TEST_691_INTERNAL2_PID || { echo "fail"; return 45; }
+
+  echo "*** HTTP server running with PID $CVMFS_TEST_691_EXTERNAL2_PID"
+
+  local external1_log="${scratch_dir}/external1.log"
+  echo "*** Start an HTTP server to serve external files (logging to $external1_log)"
+  CVMFS_TEST_691_EXTERNAL1_PID="$(open_http_server $external_storage1 $external_port1 $external1_log)"
+  [ ! -z $CVMFS_TEST_691_EXTERNAL1_PID ] && kill -0 $CVMFS_TEST_691_EXTERNAL1_PID || { echo "fail"; return 46; }
+  echo "*** HTTP server running with PID $CVMFS_TEST_691_EXTERNAL1_PID"
+
+  local external2_log="${scratch_dir}/external2.log"
+  echo "*** Start another HTTP server to serve external files (logging to $external2_log)"
+  CVMFS_TEST_691_EXTERNAL2_PID="$(open_http_server $external_storage2 $external_port2 $external2_log)"
+  [ ! -z $CVMFS_TEST_691_EXTERNAL2_PID ] && kill -0 $CVMFS_TEST_691_EXTERNAL2_PID || { echo "fail"; return 47; }
+  echo "*** HTTP server running with PID $CVMFS_TEST_691_EXTERNAL2_PID"
+
+  local metalink_internal_log="${scratch_dir}/metalink_internal.log"
+  echo "*** Start internal Metalink server (logging to $metalink_internal_log)"
+  CVMFS_TEST_691_METALINK1_PID="$(open_metalink_server "$internal_http_base1 $internal_http_base2" $metalink_internal_port $metalink_internal_log)"
+  [ ! -z $CVMFS_TEST_691_METALINK1_PID ] && kill -0 $CVMFS_TEST_691_METALINK1_PID || { echo "fail"; return 48; }
+  echo "*** Metalink server running with PID $CVMFS_TEST_691_METALINK1_PID"
+
+  local metalink_external_log="${scratch_dir}/metalink_external.log"
+  echo "*** Start external Metalink server (logging to $metalink_external_log)"
+  CVMFS_TEST_691_METALINK2_PID="$(open_metalink_server "-r $external_http_base2 $external_http_base1" $metalink_external_port $metalink_external_log)"
+  [ ! -z $CVMFS_TEST_691_METALINK2_PID ] && kill -0 $CVMFS_TEST_691_METALINK2_PID || { echo "fail"; return 49; }
+  echo "*** Metalink server running with PID $CVMFS_TEST_691_METALINK2_PID"
+
+  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+  local test_pipe=$(cat /etc/cvmfs/repositories.d/$CVMFS_TEST_REPO/client.conf | \
+                      grep ^CVMFS_TALK_SOCKET= | cut -d= -f2)
+
+  echo "*** Read first file from first internal server"
+  cat ${repo_dir}/internal/file1                             || return 51
+  [ "$(attr -qg host $cvmfs_mnt)" = "$internal_http_base1" ] || return 52
+
+  echo "*** Try reading second internal file, should fail"
+  ! cat ${repo_dir}/internal/file2                           || return 53
```

Comment:
There's a test failure here, in fact I'm not sure why this should fail. We are failing over to the right host, no?

Reply:
Hmm, it didn't in my testing.  I will need to investigate and probably think of another way to verify this. Actually I'm pretty sure I did see it failover at one point but later it didn't happen for me.

In the early days of cvmfs I was told that the client expected all files in a catalog to be on the server that it got the catalog from and otherwise it would be a fatal error, and so it was important to not use a round-robin or move a virtual IP between two individually updated stratum 1s.  It was the motivation for the creation of the [cvmfs-hastratum1](https://github.com/cvmfs-contrib/cvmfs-hastratum1) package.  Maybe that's not still the case.

Reply:
In my testing, the output from the cat command is
```
cat: /cvmfs/test.cern.ch/internal/file2: Input/output error
```
and /var/log/messages contains
```
Nov  8 12:32:45 fermicloud731 cvmfs2[1840709]: (test.cern.ch) failed to fetch /internal/file2 (hash: 1fd57eb6c654a4245aadb43480ddf64c85cfa55e, error 9 [host returned HTTP error])
Nov  8 12:32:45 fermicloud731 cvmfs2[1840709]: (test.cern.ch) failed to open inode: 262, CAS key 1fd57eb6c654a4245aadb43480ddf64c85cfa55e, error code 2
Nov  8 12:32:45 fermicloud731 cvmfs2[1840709]: (test.cern.ch) EIO (06) on /internal/file2
```
and the debug log shows
```
(download) (manager standard - id 0) reading from host 0    [11-08-2024 12:32:45 CST]
(download) (id 0) escaped http://localhost/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e to http://localhost/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e    [11-08-2024 12:32:45 CST]
(curl) (id 0) {header/sent} GET /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e HTTP/1.1
Host: localhost
Accept: */*
Connection: Keep-Alive
User-Agent: cvmfs Fuse 2.12.0    [11-08-2024 12:32:45 CST]
(curl) (id 0) {header/recv} HTTP/1.1 404 Not Found    [11-08-2024 12:32:45 CST]
(download) (id 0) http status error code: HTTP/1.1 404 Not Found
 [404]    [11-08-2024 12:32:45 CST]
(curl) (id 0) {info} Failed writing header    [11-08-2024 12:32:45 CST]
(download) (manager 'standard' - id 0) Number of CURL redirects 0    [11-08-2024 12:32:45 CST]
(download) (manager 'standard' - id 0) Verify downloaded url /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e, proxy DIRECT (curl error 23)    [11-08-2024 12:32:45 CST]
(download) (manager 'standard' - id 0) download failed (error 9 - host returned HTTP error)    [11-08-2024 12:32:45 CST]
(cache) failed to fetch /internal/file2 (hash: 1fd57eb6c654a4245aadb43480ddf64c85cfa55e, error 9 [host returned HTTP error])    [11-08-2024 12:32:45 CST]
(cache) abort ./txn/fetch9vNVaL    [11-08-2024 12:32:45 CST]
(cvmfs) failed to open inode: 262, CAS key 1fd57eb6c654a4245aadb43480ddf64c85cfa55e, error code 2    [11-08-2024 12:32:45 CST]
```

Where does the output of the 691-metalink test log show up?  I expected it to be in the test archive logs but I couldn't find them.

Reply:
Yes, for github actions the test log is stored as an artifact, that can be accessed via the "Summary" button. I'll link it when it finishes, but it anyway won't have the debuglog.

In my local test I get something like this:

```
(cache) miss: /internal/file2 /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e    [11-08-2024 17:38:17 CET]
2974 (download) (manager standard - id 1) reading from host 0    [11-08-2024 17:38:17 CET]
2975 (download) (id 1) escaped http://localhost/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e to http://localhost/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e    [11-08-2024 17:38:17 CET]
2976 (curl) (id 1) {header/sent} GET /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e HTTP/1.1^M
2977 Host: localhost^M
2978 Accept: */*^M
2979 Connection: Keep-Alive^M
2980 User-Agent: cvmfs Fuse 2.12.0    [11-08-2024 17:38:17 CET]
2981 (curl) (id 1) {header/recv} HTTP/1.1 404 Not Found    [11-08-2024 17:38:17 CET]
2982 (download) (id 1) http status error code: HTTP/1.1 404 Not Found^M
2983  [404]    [11-08-2024 17:38:17 CET]
2984 (curl) (id 1) {info} Failed writing header    [11-08-2024 17:38:17 CET]
2985 (download) (manager 'standard' - id 1) Number of CURL redirects 0    [11-08-2024 17:38:17 CET]
2986 (download) (manager 'standard' - id 1) Verify downloaded url /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e, proxy DIRECT (curl error 23)    [11-08-2024 17:38:17 CET]
2987 (download) (manager 'standard' - id 1) Trying again on same curl handle, same url: 0, error code 9 no-cache 0    [11-08-2024 17:38:17 CET]
2988 (download)  - id 1) switching host from http://localhost to http://localhost:11691 (host returned HTTP error)    [11-08-2024 17:38:17 CET]
2989 (download) (manager standard - id 1) reading from host 1    [11-08-2024 17:38:17 CET]
2990 (download) (id 1) escaped http://localhost:11691/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e to http://localhost:11691/data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e    [11-08-2024 17:38:17 CET]
2991 (curl) (id 1) {header/sent} GET /data/1f/d57eb6c654a4245aadb43480ddf64c85cfa55e HTTP/1.1^M
2992 Host: localhost:11691^M
2993 Accept: */*^M
2994 Connection: Keep-Alive^M
2995 User-Agent: cvmfs Fuse 2.12.0    [11-08-2024 17:38:17 CET]
2996 (curl) (id 1) {header/recv} HTTP/1.0 200 OK    [11-08-2024 17:38:17 CET] 
2997 (curl) (id 1) {header/recv} Server: HTTPRangeServer/0.2 Python/2.7.18    [11-08-2024 17:38:17 CET]
2998 (curl) (id 1) {header/recv} Date: Fri, 08 Nov 2024 16:38:17 GMT    [11-08-2024 17:38:17 CET]
2999 (curl) (id 1) {header/recv} Content-Type: application/octet-stream    [11-08-2024 17:38:17 CET]
3000 (curl) (id 1) {header/recv} Content-Length: 11    [11-08-2024 17:38:17 CET]
```